spec.md

```diff
@@ -3767,6 +3912,282 @@ x.remove(2)                             # None (x == [1, 3])
+while the `difference_update` method accepts arbitrary iterable sequences.
+
+It is permissible to call `difference_update` without any arguments; this leaves
+the set `S` unchanged.
+
+`difference_update` fails if the set `S` is frozen or has active iterators.
+
+```python
+s = set([1, 2, 3, 4])
+s.difference_update([2])             # None; s is set([1, 3, 4])
+s.difference_update([0, 1], [4, 5])  # None; s is set([3])
+```
+
+<a id='set·discard'></a>
+
+### set·discard
+
+`S.discard(x)` removes the value `x` from the set `S` if present. It returns
+`None`.
+
+It is permissible to `discard` a value not present in the set; this leaves the
+set `S` unchanged. If you want to fail on an attempt to remove a non-present
+element, use [`remove`](#set·remove) instead. If you need to remove multiple
+elements from a set, see [`difference_update`](#set·difference_update) or the
+[`-=`](#sets) augmented assignment operation.
+
+`discard` fails if the set `S` is frozen or has active iterators, even if `x` is
+not a member of the set.
+
+```python
+s = set(["x", "y"])
+s.discard("y")  # None; s == set(["x"])
+s.discard("y")  # None; s == set(["x"])
+```
+
+<a id='set·intersection'></a>
+
+### set·intersection
+
+`S.intersection(*others)` returns a new set containing those elements that the
+set `S` and all of the iterable sequences `*others` have in common.
+
+If `s` and `t` are sets, `s.intersection(t)` is equivalent to `s & t`; however,
+note that the `&` operation requires both sides to be sets, while the
+`intersection` method accepts arbitrary iterable sequences.
+
+It is permissible to call `intersection` without any arguments; this returns a
+copy of the set `S`.
+
+```python
+set([1, 2]).intersection([2, 3])             # set([2])
+set([1, 2, 3]).intersection([0, 1], [1, 2])  # set([1])
+```
+
+<a id='set·intersection_update'></a>
+
+### set·intersection\_update
+
+`S.intersection_update(*others)` removes from the set `S` any elements not found
+in at least one of the iterable sequences `*others`. It returns `None`.
+
+If `s` and `t` are sets, `s.intersection_update(t)` is equivalent to `s &= t`;
+however, note that the `&=` augmented assignment requires both sides to be sets,
+while the `intersection_update` method accepts arbitrary iterable sequences.
+
+It is permissible to call `intersection_update` without any arguments; this
+leaves the set `S` unchanged.
+
+`intersection_update` fails if the set `S` is frozen or has active iterators.
+
+```python
+s = set([1, 2, 3, 4])
+s.intersection_update([0, 1, 2])       # None; s is set([1, 2])
+s.intersection_update([0, 1], [1, 2])  # None; s is set([1])
+```
+
+<a id='set·isdisjoint'></a>
+
+### set·isdisjoint
+
+`S.isdisjoint(x)` returns `True` if the set `S` and the iterable sequence `x` do
+not have any values in common, and `False` otherwise.
+
+This is equivalent to `not S.intersection(x)`.
+
+`isdisjoint` fails if `x` is not an iterable sequence.
```

Comment:
We could also say "or contains unhashable elements". (Same might be said in other locations as well, but for many of the operations you implicitly guard against this when you say the semantics are the same as `s += t` etc.)

Reply:
No, we only need to check hashability of an element if we are *adding that element to a set* (such as in `union`, `update`, `symmetric_difference`, or `symmetric_difference_update`).

If we could not possibly be adding an element to a set, but are merely checking for its presence in a set (as in `isdisjoint`, `issubset`, `issuperset`, `intersection`, or `difference`) or removing it if it is present (as in `difference_update` and `intersection_update`), failing on unhashable values is pointless because an unhashable value cannot possibly be an element of a set.

Reply:
Again, precedent from Python:

```
>>> {1}.isdisjoint([2, []])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unhashable type: 'list'
```

I didn't test all the operations, and if you told me that Python is inconsistent, I wouldn't doubt you for a moment. If in fact Python *is* inconsistent then we have a freer hand to deviate from its behavior and pick what we think best. If it's uniform then there's a more compelling motivation to stay consistent with it.

I don't have a super strong opinion either way, but maybe an argument can be better phrased in terms of least-surprise or failing-fast. Or maybe there are existing operations in Starlark that can't possibly have an effect but still violate a precondition. Couldn't the same argument you're using also justify allowing a no-op mutation on a frozen set, like `my_frozen_set.update([])`?

Reply:
Thanks, fixed now - added requirement for the argument to discard (and all other set methods) to not contain unhashable values.